src/EFCore.Relational/Query/Internal/CaseSimplifyingExpressionVisitor.cs

```diff
@@ -0,0 +1,219 @@
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+        /// </summary>
+        protected override Expression VisitExtension(Expression extensionExpression)
+        {
+            Check.NotNull(extensionExpression, nameof(extensionExpression));
+
+            if (extensionExpression is ShapedQueryExpression shapedQueryExpression)
+            {
+                return shapedQueryExpression.Update(Visit(shapedQueryExpression.QueryExpression), shapedQueryExpression.ShaperExpression);
+            }
+
+            // Only applies to 'CASE WHEN condition...' not 'CASE operand WHEN...'
+            if (extensionExpression is CaseExpression caseExpression
+                && caseExpression.Operand == null
+                && caseExpression.ElseResult is CaseExpression nestedCaseExpression
+                && nestedCaseExpression.Operand == null)
+            {
+                return VisitExtension(_sqlExpressionFactory.Case(
+                    caseExpression.WhenClauses.Union(nestedCaseExpression.WhenClauses).ToList(),
+                    nestedCaseExpression.ElseResult));
+            }
+
+            if (extensionExpression is SqlBinaryExpression sqlBinaryExpression)
+            {
+                var sqlConstantComponent = sqlBinaryExpression.Left as SqlConstantExpression ?? sqlBinaryExpression.Right as SqlConstantExpression;
+                var caseComponent = sqlBinaryExpression.Left as CaseExpression ?? sqlBinaryExpression.Right as CaseExpression;
+
+                // generic CASE statement comparison optimization:
+                // (CASE
+                //  WHEN condition1 THEN result1
+                //  WHEN condition2 THEN result2
+                //  WHEN ...
+                //  WHEN conditionN THEN resultN) == result1 -> condition1
+                if (sqlBinaryExpression.OperatorType == ExpressionType.Equal
+                    && sqlConstantComponent != null
+                    && sqlConstantComponent.Value != null
+                    && caseComponent != null
+                    && caseComponent.Operand == null)
+                {
+                    var matchingCaseBlock = caseComponent.WhenClauses.FirstOrDefault(wc => sqlConstantComponent.Equals(wc.Result));
+                    if (matchingCaseBlock != null)
+                    {
+                        return matchingCaseBlock.Test;
+                    }
+                }
+
+                // CompareTo specific optimizations
+                if (sqlConstantComponent != null
+                    && sqlConstantComponent.Value != null
+                    && IsCompareTo(caseComponent)
+                    && sqlConstantComponent.Value is int intValue
+                    && (intValue == 0 || intValue == 1 || intValue == -1)
```

Comment:
```suggestion
                    && (intValue > -2 && intValue < 2)
```
could be faster. :trollface: